components/script/dom/bindings/str.rs

```diff
@@ -208,6 +208,114 @@ impl DOMString {
         self.0.truncate(last_non_whitespace);
         let _ = self.0.splice(0..first_non_whitespace, "");
     }
+
+    /// A valid date string should be YYYY-MM-DD
+    /// YYYY must be four or more digits, MM and DD both must be two digits
+    /// https://html.spec.whatwg.org/multipage/#valid-date-string
+    pub fn is_valid_date_string(&mut self) -> bool {
+        match self.parse_date_component(&*self.0) {
+            Ok(_)  => true,
+            Err(_) => false
+        }
+    }
+
+    /// A valid month string should be "YYYY-MM"
+    /// YYYY must be four or more digits, MM both must be two digits
+    /// https://html.spec.whatwg.org/multipage/#valid-month-string
+    pub fn is_valid_month_string(&mut self) -> bool {
+        match self.parse_month_component(&*self.0) {
+            Ok(_)  => true,
+            Err(_) => false
+        }
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#parse-a-month-component
+    fn parse_month_component(&self, value: &str) -> Result<(u32, u32), ()> {
+        // Step 3
+        let value_vec: Vec<&str> = value.split('-').collect();
+        if value_vec.len() < 2 || value_vec.len() > 3 {
+            return Err(());
+        }
+        let year = value_vec[0];
+        let month = value_vec[1];
+
+        // Step 1, 2
+        if year.len() < 4 ||
+           !year.chars().all(|c| c.is_digit(10)) ||
+           year == "0000" {
+            return Err(());
+        }
+
+        // Step 4, 5
+        let month_int = month.parse::<u32>().unwrap();
+        if !(month.len() == 2 && month.chars().all(|c| c.is_digit(10))) {
```

Comment:
The check for a base-10 digit is unnecessary here, since the parse method call above would've returned an Err if any of the characters is not base 10.